include/exadg/poisson/spatial_discretization/operator.cpp

```diff
@@ -135,57 +136,44 @@ Operator<dim, n_components, Number>::distribute_dofs()
 
   dof_handler.distribute_dofs(*fe);
 
-  // Affine constraints are only relevant for continuous Galerin discretization.
-  // The AffineConstraints object is used to initialize MatrixFree. Here, we apply homogeneous
-  // boundary conditions as needed by vmult() in iterative solvers for linear systems of equations,
-  // implemented via dealii::MatrixFree and FEEvaluation::read_dof_values() (or gather_evaluate()).
-  // The actual inhomogeneous boundary data needs to be imposed separately due to the implementation
-  // in deal.II that can not handle multiple AffineConstraints. Hence, we need to do this explicitly
-  // in ExaDG.
+  // Affine constraints are only relevant for continuous Galerkin discretizations.
   if(param.spatial_discretization == SpatialDiscretization::CG)
   {
+    affine_constraints_periodicity_and_hanging_nodes.clear();
+
+    add_hanging_node_and_periodicity_constraints(affine_constraints_periodicity_and_hanging_nodes,
+                                                 *this->grid,
+                                                 dof_handler);
+
+    affine_constraints_periodicity_and_hanging_nodes.close();
```

Comment:
this line seems to have introduced a bug in debug mode, see PR #535, since we later call `copy_from` and then add additional constraints.

Reply:
The solution is to call `close()` in the end once all the constraints have been added (which somewhat hinders writing modular code ...)